runtime/src/iree/hal/drivers/hip/graph_command_buffer.c

```diff
@@ -62,15 +67,95 @@ typedef struct iree_hal_hip_graph_command_buffer_t {
 static const iree_hal_command_buffer_vtable_t
     iree_hal_hip_graph_command_buffer_vtable;
 
+static iree_status_t
+iree_hal_hip_graph_command_buffer_execution_barrier_internal(
+    iree_hal_hip_graph_command_buffer_t* command_buffer);
+
 static iree_hal_hip_graph_command_buffer_t*
 iree_hal_hip_graph_command_buffer_cast(iree_hal_command_buffer_t* base_value) {
   IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_graph_command_buffer_vtable);
   return (iree_hal_hip_graph_command_buffer_t*)base_value;
 }
 
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION_DEVICE
+
+static void iree_hip_graph_command_buffer_trace_zone_begin_external(
+    iree_hal_hip_graph_command_buffer_t* command_buffer, const char* file_name,
+    size_t file_name_length, uint32_t line, const char* function_name,
+    size_t function_name_length, const char* name, size_t name_length) {
+  // Make sure there are no new nodes after the last barrier.
+  // Work should start after the event.
+  if (IREE_UNLIKELY(command_buffer->graph_node_count != 0)) {
+    iree_hal_hip_graph_command_buffer_execution_barrier_internal(
+        command_buffer);
+  }
+
+  hipGraphNode_t* tracing_event_node =
+      &command_buffer->hip_graph_nodes[command_buffer->graph_node_count++];
+  size_t dependency_count = command_buffer->hip_barrier_node ? 1 : 0;
+  IREE_HIP_GRAPH_TRACE_ZONE_BEGIN_EXTERNAL(
+      command_buffer->tracing_context, tracing_event_node,
+      command_buffer->hip_graph, &command_buffer->hip_barrier_node,
+      dependency_count, file_name, file_name_length, line, function_name,
+      function_name_length, name, name_length);
+
+  // Move the barrier forward to make sure that the tracing event is recorded
+  // before work starts.
+  // Downstream operations will wait on the tracing node.
+  command_buffer->hip_barrier_node = *tracing_event_node;
+}
+
+static void iree_hip_graph_command_buffer_trace_zone_end(
+    iree_hal_hip_graph_command_buffer_t* command_buffer) {
+  // Make sure there are no new nodes after the last barrier.
+  // Prior work should end before the tracing event is recorded.
+  if (IREE_UNLIKELY(command_buffer->graph_node_count != 0)) {
+    iree_hal_hip_graph_command_buffer_execution_barrier_internal(
+        command_buffer);
+  }
+
+  hipGraphNode_t* tracing_event_node =
+      &command_buffer->hip_graph_nodes[command_buffer->graph_node_count++];
+  size_t dependency_count = command_buffer->hip_barrier_node ? 1 : 0;
+  IREE_ASSERT(dependency_count > 0 &&
+              "Ending a zone should at least depend on the beginning.");
```

Comment:
Use `IREE_ASSERT_GT` and for the error message avoid captialize first char and avoid the end `.`

Reply:
done